### DISCUSSION TITLE
Should I use IJsInProcessRuntime or [JSImport] API for "regular javascript interop calls?

### DISCUSSION THREAD

**Discussion by Matheos96** [dotnet/aspnetcore#53650] [Q&A] [answered]
We just migrated our .NET 6 code base to .NET 8 and I am familiarizing myself with what we can do in blazor with the new additions from .NET 7 and 8. We are running Blazor WASM Standalone btw.

Earlier we used `IJsInProcessRuntime `and `IJsUnmarshalledRuntime `(or whatever it is called). The latter is now deprecated in .NET 7 and we have been blessed with the new` [JSImport]`/`[JSExport]` API. My question is, since `IJSInProcessRuntime `is not deprecated, which one is the better option, `JSImport `or `IJsInProcessRuntime`? Performance wise. One thing I have noticed is that I don't think we can call functions on JS classes using `JSImport`, like we can with `IJsInProcessRuntime `and a `JSObjectReference`.

Again, we are on WASM only, so we can use sync only calls as much as possible

**Comment by javiercn** [comment, marked as answer]
IJsInProcessRuntime is the only thing officially supported by Blazor. If you rely on `[JSImport]` and `[JSExport]` it's not guaranteed that some Blazor features work. That said, if you want the maximum performance (in 99% of the cases you don't need it) `[JSImport]` and `[JSExport]` are the way to go.

**Reply by Matheos96** [reply]
Okay, we are actually quite interested in improving performance and were already using the old JSUnmarshalled earlier before it was deprecated. What do you mean by it not being officially supported? I feel like even the examples with blazor code?

**Reply by javiercn** [reply]
I don't know for sure, but likely the answer is no, features like `IJSObjectReference` won't work with JSImport/Export

**Reply by javiercn** [reply]
And that [JSImport] and [JSExport] won't work with Blazor Server.

**Reply by Matheos96** [reply]
Okay yes, then we are on the same page. We are on wasm only so that is no issue, but I was wondering a bit regarding IJSobjectreference. So at least I cannot call "instance" methods using a IJSObjectreference and JSImport, but can I pass the object itself through JSImport and call methods on the JS side?

**Reply by javiercn** [reply]
> So at least I cannot call "instance" methods using a IJSObjectreference and JSImport, but can I pass the object itself through JSImport and call methods on the JS side?

I don't believe so. I think that [JSImport] and [JSExport] have their own way of achieving that.

**Reply by Matheos96** [reply]
Okay, I guess I will just have to give it a try. Thanks for the info! The documentation overall regarding the JSImport stuff is not too well covered in my opinion. Yes, we have the official docs but it is not too in depth with actual use case scenarios.
